File: framework-cs/apps/UnityExample/Assets/Samples/Devian Foundation/0.1.0/GamePackage/Runtime/Ability/AbilityBattleProjectile.cs
using System;
using System.Collections.Generic;
using Devian.Domain.Game;

namespace Devian
{
    public sealed class AbilityBattleProjectile : AbilityBattleBase
    {
        PROJECTILE mTable = null;

        public string ProjectileId => mTable?.projectile_id ?? string.Empty;
        public string NameId => mTable?.name_id ?? string.Empty;
        public IReadOnlyList<string> AffectList => mTable != null ? mTable.affect_list : Array.Empty<string>();

        public void Init(PROJECTILE table)
        {
            mTable = table;
        }

        public override AbilityBase Clone()
        {
            var c = new AbilityBattleProjectile();
            c.mTable = mTable;
            c.CopyStatsFrom(this);
            return c;
        }
    }
}

File: framework-ts/module/devian-domain-game/features/ability/AbilityBattleProjectile.ts
import { PROJECTILE } from '../../Generated/Game.g';
import { AbilityBattleBase } from './AbilityBattleBase';

export class AbilityBattleProjectile extends AbilityBattleBase {
    private mTable: PROJECTILE | null = null;

    get projectileId(): string { return this.mTable?.projectile_id ?? ''; }
    get nameId(): string { return this.mTable?.name_id ?? ''; }
    get affectList(): readonly string[] { return this.mTable?.affect_list ?? []; }

    init(table: PROJECTILE): void {
        this.mTable = table;
    }

    clone(): AbilityBattleProjectile {
        const c = new AbilityBattleProjectile();
        c.mTable = this.mTable;
        c.copyStatsFrom(this);
        return c;
    }
}
